src/sparse/impl/KokkosSparse_spmv_bsrmatrix_impl.hpp

```diff
@@ -608,58 +613,69 @@ struct BSR_GEMV_Functor {
         KokkosBatched::SerialGemvInternal<KokkosBatched::Algo::Gemv::Blocked>::
-            invoke<value_type, value_type>(block_dim, block_dim, alpha,
-                                           Aview.data(), block_dim, 1,
-                                           &m_x(xstart), m_x.stride_0(), beta1,
-                                           &m_y(ystart), m_y.stride_0());
+            invoke<value_type, value_type>(
+                static_cast<int>(block_dim), static_cast<int>(block_dim), alpha,
+                Aview.data(), static_cast<int>(block_dim), 1, &m_x(xstart),
+                static_cast<int>(m_x.stride_0()), beta1, &m_y(ystart),
+                static_cast<int>(m_y.stride_0()));
       }
     }
   }
 
   KOKKOS_INLINE_FUNCTION
   void operator()(const team_member &dev) const {
-    using y_value_type = typename YVector::non_const_value_type;
-    Kokkos::parallel_for(
-        Kokkos::TeamThreadRange(dev, 0, blocks_per_team),
-        [&](const ordinal_type &loop) {
-          const ordinal_type iBlock =
-              static_cast<ordinal_type>(dev.league_rank()) * blocks_per_team +
-              loop;
-          if (iBlock >= m_A.numRows()) {
-            return;
-          }
-          const auto start = m_A.graph.row_map(iBlock);
-          const ordinal_type count =
-              static_cast<ordinal_type>(m_A.graph.row_map(iBlock + 1) - start);
-          const auto row = m_A.block_row_Const(iBlock);
-          //
-          auto yview = Kokkos::subview(
-              m_y, Kokkos::make_pair(iBlock * block_dim,
-                                     iBlock * block_dim + block_dim));
-          //
-          for (ordinal_type ir = 0; ir < block_dim; ++ir) {
-            y_value_type sum = 0;
-
-            Kokkos::parallel_reduce(
-                Kokkos::ThreadVectorRange(dev, count),
-                [&](const ordinal_type &iEntry, y_value_type &lsum) {
-                  const auto start_col = row.block_colidx(iEntry) * block_dim;
-                  for (ordinal_type jr = 0; jr < block_dim; ++jr) {
-                    const value_type val =
-                        conjugate
-                            ? ATV::conj(row.local_block_value(iEntry, ir, jr))
-                            : row.local_block_value(iEntry, ir, jr);
-                    lsum += val * m_x(start_col + jr);
-                  }
-                },
-                sum);
-
-            Kokkos::single(Kokkos::PerThread(dev), [&]() {
-              sum *= alpha;
-              yview(ir) += sum;
-            });
-          }
-        });
+    using y_value_type        = typename YVector::non_const_value_type;
+    const ordinal_type iBlock = static_cast<ordinal_type>(dev.league_rank());
+
+    const size_type Y_ptBeg = iBlock * block_dim;
+    const size_type Y_ptEnd = Y_ptBeg + block_dim;
+    auto Y_cur = Kokkos::subview(m_y, ::Kokkos::make_pair(Y_ptBeg, Y_ptEnd));
+
+    const y_value_type val_one = 1;
```

Comment:
It would probably be safer to use `Kokkos::ArithTraits<y_value_type>::one();` here

Reply:
Done.